### PR TITLE
WFLY-9842 Enhance CommandDispatcher async methods to use CompletableFutures

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.jboss.org/browse/WFLY-9842
https://issues.jboss.org/browse/EAP7-989

### CODE REVIEW

File: clustering/api/WFLY-9842_CommandDispatcher_Enhancements.adoc [hunk 1]
```diff
@@ -0,0 +1,62 @@
+= Title
```
Comment (jamezp) [inline clustering/api/WFLY-9842_CommandDispatcher_Enhancements.adoc]: This should be replaced with a title for the feature.

Comment (pferraro) [reply clustering/api/WFLY-9842_CommandDispatcher_Enhancements.adoc]: Done

Comment (jamezp) [reply clustering/api/WFLY-9842_CommandDispatcher_Enhancements.adoc]: Did you forget to push the change?

Comment (pferraro) [reply clustering/api/WFLY-9842_CommandDispatcher_Enhancements.adoc]: I did indeed. :)

**Comment by pferraro** [comment]
@tommaso-borgato No, I mean the integration tests here:
https://github.com/pferraro/wildfly/tree/server/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/dispatcher
https://github.com/pferraro/wildfly/tree/server/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/single/dispatcher

Compatibility tests for the deprecated API are in the form of simple unit tests:
https://github.com/pferraro/wildfly/blob/server/clustering/api/src/test/java/org/wildfly/clustering/dispatcher/CommandDispatcherTestCase.java

**Comment by tommaso-borgato** [comment]
@pferraro thank you. 

### Test org.jboss.as.test.clustering.cluster.dispatcher.CommandDispatcherTestCase
When I run the **first** test with either the following command:
```
./integration-tests.sh test -Dts.noSmoke -Dts.clustering -Dtest=org.jboss.as.test.clustering.cluster.dispatcher.CommandDispatcherTestCase
```
or:
```
cd ./wildfly/testsuite/integration/clustering
mvn install -Dts.noSmoke -Dts.clustering -Dtest=org.jboss.as.test.clustering.cluster.dispatcher.CommandDispatcherTestCase
```
I get the following error:
```
[ERROR] test(org.jboss.as.test.clustering.cluster.dispatcher.CommandDispatcherTestCase)  Time elapsed: 20.976 s  <<< ERROR!
javax.ejb.EJBException: java.lang.ClassCastException: org.wildfly.clustering.server@13.0.0.Alpha1-SNAPSHOT//org.wildfly.clustering.server.dispatcher.NoSuchService cannot be cast to java.base/java.lang.String
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.tx.CMTTxInterceptor.invokeInOurTx(CMTTxInterceptor.java:246)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.tx.CMTTxInterceptor.required(CMTTxInterceptor.java:362)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.tx.CMTTxInterceptor.processInvocation(CMTTxInterceptor.java:144)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext$Invocation.proceed(InterceptorContext.java:509)
        at org.jboss.weld.core@3.0.3.Final//org.jboss.weld.module.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:81)
        at org.jboss.as.weld.common@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.weld.ejb.EjbRequestScopeActivationInterceptor.processInvocation(EjbRequestScopeActivationInterceptor.java:89)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.component.interceptors.CurrentInvocationContextInterceptor.processInvocation(CurrentInvocationContextInterceptor.java:41)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.component.invocationmetrics.WaitTimeInterceptor.processInvocation(WaitTimeInterceptor.java:47)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.security.SecurityContextInterceptor.processInvocation(SecurityContextInterceptor.java:100)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.deployment.processors.StartupAwaitInterceptor.processInvocation(StartupAwaitInterceptor.java:22)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.component.interceptors.ShutDownInterceptorFactory$1.processInvocation(ShutDownInterceptorFactory.java:64)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.deployment.processors.EjbSuspendInterceptor.processInvocation(EjbSuspendInterceptor.java:57)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.component.interceptors.LoggingInterceptor.processInvocation(LoggingInterceptor.java:67)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.ee@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ee.component.NamespaceContextInterceptor.processInvocation(NamespaceContextInterceptor.java:50)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.component.interceptors.AdditionalSetupInterceptor.processInvocation(AdditionalSetupInterceptor.java:54)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.ContextClassLoaderInterceptor.processInvocation(ContextClassLoaderInterceptor.java:60)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.run(InterceptorContext.java:438)
        at org.wildfly.security.elytron-private@1.2.4.Final//org.wildfly.security.manager.WildFlySecurityManager.doChecked(WildFlySecurityManager.java:619)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.AccessCheckingInterceptor.processInvocation(AccessCheckingInterceptor.java:57)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.ChainedInterceptor.processInvocation(ChainedInterceptor.java:53)
        at org.jboss.as.ee@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ee.component.ViewService$View.invoke(ViewService.java:198)
        at org.wildfly.security.elytron-private@1.2.4.Final//org.wildfly.security.auth.server.SecurityIdentity.runAsFunctionEx(SecurityIdentity.java:380)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.remote.AssociationImpl.invokeWithIdentity(AssociationImpl.java:541)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.remote.AssociationImpl.invokeMethod(AssociationImpl.java:536)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.remote.AssociationImpl.lambda$receiveInvocationRequest$0(AssociationImpl.java:194)
        at org.jboss.threads@2.3.2.Final//org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads@2.3.2.Final//org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1985)
        at org.jboss.threads@2.3.2.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1487)
        at org.jboss.threads@2.3.2.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1378)
        at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.ClassCastException: org.wildfly.clustering.server@13.0.0.Alpha1-SNAPSHOT//org.wildfly.clustering.server.dispatcher.NoSuchService cannot be cast to java.base/java.lang.String
        at deployment.CommandDispatcherTestCase.jar//org.jboss.as.test.clustering.cluster.dispatcher.bean.ClusterTopologyRetrieverBean.getClusterTopology(ClusterTopologyRetrieverBean.java:33)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.jboss.as.ee@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ee.component.ManagedReferenceMethodInterceptor.processInvocation(ManagedReferenceMethodInterceptor.java:52)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext$Invocation.proceed(InterceptorContext.java:509)
        at org.jboss.as.weld.common@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.weld.interceptors.Jsr299BindingsInterceptor.doMethodInterception(Jsr299BindingsInterceptor.java:90)
        at org.jboss.as.weld.common@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.weld.interceptors.Jsr299BindingsInterceptor.processInvocation(Jsr299BindingsInterceptor.java:101)
        at org.jboss.as.ee@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ee.component.interceptors.UserInterceptorFactory$1.processInvocation(UserInterceptorFactory.java:63)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.component.invocationmetrics.ExecutionTimeInterceptor.processInvocation(ExecutionTimeInterceptor.java:43)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.jpa@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.jpa.interceptor.SBInvocationInterceptor.processInvocation(SBInvocationInterceptor.java:47)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.ee@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ee.concurrent.ConcurrentContextInterceptor.processInvocation(ConcurrentContextInterceptor.java:45)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InitialInterceptor.processInvocation(InitialInterceptor.java:40)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.ChainedInterceptor.processInvocation(ChainedInterceptor.java:53)
        at org.jboss.as.ee@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ee.component.interceptors.ComponentDispatcherInterceptor.processInvocation(ComponentDispatcherInterceptor.java:52)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.component.pool.PooledInstanceInterceptor.processInvocation(PooledInstanceInterceptor.java:51)
        at org.jboss.invocation@1.5.1.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
        at org.jboss.as.ejb3@13.0.0.Alpha1-SNAPSHOT//org.jboss.as.ejb3.tx.CMTTxInterceptor.invokeInOurTx(CMTTxInterceptor.java:237)
        ... 42 more

```

### Test org.jboss.as.test.clustering.single.dispatcher.CommandDispatcherTestCase
When I run the **second** test with either the following command:
```
./integration-tests.sh test -Dts.noSmoke -Dts.clustering -Dtest=org.jboss.as.test.clustering.single.dispatcher.CommandDispatcherTestCase
```
or:
```
cd ./wildfly/testsuite/integration/clustering
mvn install -Dts.noSmoke -Dts.clustering -Dtest=org.jboss.as.test.clustering.single.dispatcher.CommandDispatcherTestCase
```
I get the following error:
```
[ERROR] org.jboss.as.test.clustering.single.dispatcher.CommandDispatcherTestCase  Time elapsed: 0.003 s  <<< ERROR!
org.jboss.arquillian.container.test.impl.client.deployment.ValidationException: 
DeploymentScenario contains a target (_DEFAULT_) not matching any defined Container in the registry.
Possible causes are: None of the 5 Containers are marked as default or you have defined a @Deployment with a @TargetsContainer of value (_DEFAULT_) that does not match any found/configured Containers (node-1,node-2,node-3,node-4,node-non-ha), see arquillian.xml container@qualifier
```


Am I missing something in the way I run the tests?

**Comment by pferraro** [comment]
@tommaso-borgato Can we discuss this offline?  The CI test runs have already validated that the complete set of integration tests pass.  Let's not obfuscate these results with test runs from your own environment.

**Pull request merged**